### ISSUE TITLE
BlackBox: Internal flash memory W25N01G not supported

### ISSUE THREAD

**Issue by robustini** [iNavFlight/inav#8098]
## Current Behavior

With amazement I noticed that on my ZEEZF7 V1.2 the 128 Mb internal flash for the BlackBox is not recognized.
With Betaflight, as you can see from the image below, no problem.

## Steps to Reproduce

1. Flash iNav V5.0 RC2 or 4.1, is indifferent
2. Connect to the FC, choose the default settings and go to the BlackBox section, you will see that it does not find the flash

## Suggested solution(s)

Is it possible that the target for the ZEEZF7 regarding the internal flash is wrong?
I checked it and the pins comparing with Betaflight seem to be declared correctly.

## Additional context

iNav: onboard flash not found.
![immagine](https://user-images.githubusercontent.com/3864712/172351086-625f418c-57b2-4b2d-accf-f415384b8061.png)

Betaflight: 128 Mb onboard flash found.
![immagine](https://user-images.githubusercontent.com/3864712/172351117-c8bd0c06-359e-41f8-af41-667ab364be25.png)

---

INAV/ZEEZF7 5.0.0 Jun  3 2022 / 21:25:20 (46afd3a3)
GCC-10.2.1 20201103 (release)

**Comment by robustini** [comment]
From what I see in the code iNav does not have support for onboard flashes other than 16 Mb, at least I see this from all the various targets and imho it is a limitation, also because replacing the onboard flash from the FC is not dramatic, but if it is not then seen from the code would be of little use.

**Comment by JulioCesarMatias** [comment]
Your flight controller flash memory is `w25n01g`, and iNav does not have drivers for it.

**Comment by robustini** [comment]
> Your flight controller flash memory is `w25n01g`, and iNav does not have drivers for it.

Thanks mate!
Theoretically it would be enough to edit the file flash_m25p16.c and add it, I'm trying but then it doesn't work.

**Comment by robustini** [comment]
Damn no, I'm looking now in Betaflight and there is just a different handling for this flash, and a specific driver.

**Comment by robustini** [comment]
Damn, I'm looking now in Betaflight and there is just a different handling for this flash, and a specific driver.
If the target is officially supported by iNav I don't understand why Blackbox support hasn't been included as well.
We have created a target for https://github.com/iNavFlight/inav/pull/7055 where instead everything works, including flash, and has not been officially inserted, precisely because the lead developers have not had the opportunity to test it personally.
However, if they tested the ZEEZF7 because they then included it in the official targets, it seems to me that they have forgotten something... :P
I really don't understand this philosophy of integration! :-)
In Ardupilot if we insert the support for a FC we make sure that everything works or we don't put it in the official targets.

**Comment by erstec** [comment]
Every development may have bugs, developers are peoples, peoples tend to make mistakes, regarding anyting.

If you just found a bug and registered it, it just a question who will take care and fix it. In this particular case FC is needed to test functionality. Probably someone from dev's have it on a bench.

It was integrated https://github.com/iNavFlight/inav/pull/5603 and https://github.com/iNavFlight/inav/pull/5787
I just wondering how it was ok for last two years, whatever.

**Comment by robustini** [comment]
Of course, mine was only a funny provocation, nothing against developers, of course!
Since we have made the official support for a target that has not been inserted then I wondered why this one that is officially supported does not work as it should.
There are perhaps too many variants of ZEEZF7, and I see only two in the target, while the hardware versions are four, I think, 1.0, 1.1, 1.2 and 2.0.
Probably between the various hardware versions they have installed different flashes, and the developers may not know this.
So if is possible let it be fixed, thank you!

**Comment by JulioCesarMatias** [comment]
> Damn, I'm looking now in Betaflight and there is just a different handling for this flash, and a specific driver. If the target is officially supported by iNav I don't understand why Blackbox support hasn't been included as well. We have created a target for [GEPRCF722](https://github.com/iNavFlight/inav/pull/7055) where instead everything works, including flash, and has not been officially inserted, precisely because the lead developers have not had the opportunity [to test it personally](https://github.com/iNavFlight/inav/blob/master/docs/policies/NEW_HARDWARE_POLICY.md). However, if they tested the ZEEZF7 because they then included it in the official targets, it seems to me that they have forgotten something... :P I really don't understand this hardware policy! :-) In Ardupilot if we insert the support for a FC we make sure that everything works or we don't put it in the official targets. Anyway @JulioCesarMatias thanks for this, I'm forced to use Betaflight because I need the flash... :-(

i will try to bring this support to iNav

**Comment by robustini** [comment]
@JulioCesarMatias knowing which ones I don't think is simple since you know better than me that the flash chip is detected by the code and then used the driver accordingly.

**Comment by JulioCesarMatias** [comment]
same problem >> https://github.com/iNavFlight/inav/issues/7885

**Comment by JulioCesarMatias** [comment]
@robustini can you test if this works?

https://drive.google.com/file/d/1peYt-UIxz3xIswWzr7wCeyvK7PyFkjY6/view?usp=sharing

**Comment by JulioCesarMatias** [comment]
Hello @robustini 

can you test if this works?

.Hex file >> https://drive.google.com/file/d/1peYt-UIxz3xIswWzr7wCeyvK7PyFkjY6/view?usp=sharing

**Comment by robustini** [comment]
> Hello @robustini
> 
> can you test if this works?
> 
> .Hex file >> https://drive.google.com/file/d/1peYt-UIxz3xIswWzr7wCeyvK7PyFkjY6/view?usp=sharing

Hello @JulioCesarMatias, thanks for trying but "Your flight controller does not have a compatible dataflash chip available".

**Comment by JulioCesarMatias** [comment]
OK thank you! I tried, but not having an FC with that flash memory in hand is hard to make it work. Here's what I tried (in case anyone wants to try to get support working) >> https://github.com/JulioCesarMatias/inav/commit/42022c34b98c75a982e9a8bdd9f8495a520ee2bc

**Comment by robustini** [comment]
> OK thank you! I tried, but not having an FC with that flash memory in hand is hard to make it work. Here's what I tried (in case anyone wants to try to get support working) >> [JulioCesarMatias@42022c3](https://github.com/JulioCesarMatias/inav/commit/42022c34b98c75a982e9a8bdd9f8495a520ee2bc)

If you tell me how I can help you I am at your disposal, I can compile the code too.

**Comment by JulioCesarMatias** [comment]
> > OK thank you! I tried, but not having an FC with that flash memory in hand is hard to make it work. Here's what I tried (in case anyone wants to try to get support working) >> [JulioCesarMatias@42022c3](https://github.com/JulioCesarMatias/inav/commit/42022c34b98c75a982e9a8bdd9f8495a520ee2bc)
> 
> If you tell me how I can help you I am at your disposal, I can compile the code too.

you can download and try to make this work from my Branch that I sent in the previous comment...

**Comment by robustini** [comment]
> ![H50bac280a215424b81617e3988af6354S](https://user-images.githubusercontent.com/25625982/175180218-fdd97f2f-5ab8-4ad7-81aa-209753ab92a9.png)
> 
> I found a ZeezF7 that uses the `25Q128` chip, are we sure your FC uses the `W25N01G`? Could you verify this?

I'm not absolutely sure if that type of dataflash is mounted, unfortunately I have the racer mounted with the board in roll180 and the chip is underneath, I can't see its label.
However even that doesn't work.
I sent an email to those who marketed it to find out which dataflash mounts but at the moment I have no answer.
Is there no way in Betaflight from the CLI to detect which one is with a command, or via a debug log?
This is taken from the pdf but it cannot be read here either.

![Cattura](https://user-images.githubusercontent.com/3864712/175231555-5bab0e6a-1fe7-43b5-83b4-b5af4b38bd69.JPG)

**Comment by robustini** [comment]
@JulioCesarMatias but I tell you this. I had tried to manually insert all the dataflashes that are in Betaflight, respecting the syntax, I compiled but it didn't work anyway, but obviously I hadn't put the W25N01G which contemplated separate libraries.
So now the curiosity to identify the correct label on the chip increases.

**Comment by robustini** [comment]
@JulioCesarMatias anyway with Betaflight I forgot to tell you that it is seen, but when i activate the MSC mode anyway Windows cannot mount the mass memory, so it is still unusable.
By activating the blackbox on ALWAYS it is seen that the free space decreases leaving the FC on, then it writes but then nothing is read, so it is seen but it is as if it were not identified... :-)

**Comment by robustini** [comment]
I'm still comparing the dataflash libraries, in Betaflight I see W25Q128_DTR and ZB25VQ128 which are not in iNav, I don't see any others that may be missing.

**Comment by robustini** [comment]
@JulioCesarMatias the author of the FC, the dearest Guido Fabbrini answered me and showed me the correct dataflash: W25N01GVZEIG

https://www.digikey.com/en/products/detail/winbond-electronics/W25N01GVZEIG-TR/5803931?fbclid=IwAR3aZ1SdJ61EL5unSt20w4R3IMFGieCX07tXPXy-YxW9-CJU1czWW8sNXA4

**Comment by hogthrob** [comment]
I look a bit into the implementation @JulioCesarMatias thankfully provided although he doesn't have a target board with said chip. There are a couple of problems there so the code cannot work without modifications. I went ahead as I am also affected (the MAMBA H743 has the same chip) and did some modifications. Now the chip is being detected correctly on my board but writing/erasing is not working (yet).  I will publish the code in my repo soon.

**Comment by hogthrob** [comment]
Here we go: https://github.com/hogthrob/inav/tree/flash_w25n01gSupport

As written in the commit comment: Reading works, only tested on MAMBA H743. However, should also work in the same (limited) way with the ZEEZF7 if pin definitions in target.h are correct for the target, i.e. it should be able to detect the chip.

Please note: Besides the code not working  fully I don't have the correct formatting settings for VS code, hence a couple more changes than really done are  indicated unfortunately.

**Comment by JulioCesarMatias** [comment]
very nice @hogthrob, when you get Read/Write working, immediately open a PR, that will be great!

**Comment by robustini** [comment]
@hogthrob thanks for the fix, i can confirm, now the dataflash is detected but maybe the writings are not correct, if I delete it and I reboot I find it completely full.
In the meantime it is a good starting point.

![immagine](https://user-images.githubusercontent.com/3864712/175820356-51b79447-da2d-4a59-bf2a-1964ec3e95f2.png)

**Comment by JulioCesarMatias** [comment]
> I couldn't find which part of the code he modified so that the flash was found

**Comment by JulioCesarMatias** [comment]
> @hogthrob thanks for the fix, i can confirm, now the dataflash is detected but maybe the writings are not correct, if I erase it and I reboot I find it completely full. In the meantime it is a good starting point.
> 
> ![immagine](https://user-images.githubusercontent.com/3864712/175820356-51b79447-da2d-4a59-bf2a-1964ec3e95f2.png)

I couldn't find which part of the code he modified so that the flash was found

**Comment by JulioCesarMatias** [comment]
> > @hogthrob thanks for the fix, i can confirm, now the dataflash is detected but maybe the writings are not correct, if I erase it and I reboot I find it completely full. In the meantime it is a good starting point.
> > ![immagine](https://user-images.githubusercontent.com/3864712/175820356-51b79447-da2d-4a59-bf2a-1964ec3e95f2.png)
> 
> I couldn't find which part of the code he modified so that the flash was found

oh ok, I could understand where my mistake was... The m25p16 flash return function did inside flashSpiInit caused the w25n01g initialization to be ignored

**Comment by hogthrob** [comment]
@JulioCesarMatias: yes, that was part of the problem. 
The other problem was/is that some required code is bound to the USE_FLASH_M25P16 so both USE_FLASH_W25N01G have to be defined, but that can be relatively easily fixed, of course. Not knowing the way how you iNav guys would like to solve this, I just left this as it was mostly by not changing anything outside `flash*` (plus of course the target.h). That can be sorted out later.

I am not clear yet why writing is not working. On first glance code looks reasonable, when comparing it to betaflight's code for doing the same job.

**Comment by hogthrob** [comment]
Ok, the (or at least one) problem is that the code to write a page only writes if it has been completely filled to the border of the next:
```C
void w25n01g_pageProgramFinish(void)
{
    if (bufferDirty && W25N01G_LINEAR_TO_COLUMN(programLoadAddress) == 0)
    {
```
In betaflight this is handled through the flush operation which presumably is called at the end of a write cycle. However, this is not happening in iNav as flush is never called. So nothing gets written out unless writing exactly on page boundaries.
In betaflight there is some higher level code in flash.c to make that work. This high level code is shared code and used for multiple low-level drivers for the various flash chip families supported in betaflight.

In a nutshell: while I have not yet verified this, I would assume that the low-level write code for the W25N01G should work. But iNav misses some extra code to handle the more complex high-level write approach required for the low-level code to do its job.

**Comment by JulioCesarMatias** [comment]
@robustini can you test again? >> https://drive.google.com/file/d/1vEfAofAj3jvgM9LFk5VcwXBSCSKxllbM/view?usp=sharing

**Comment by hogthrob** [comment]
@JulioCesarMatias ,  haven't had the opportunity to review your changes, but you may want to  have a look at my last commit https://github.com/iNavFlight/inav/commit/101e345b4f21afe2319c0235d0762082126a9d09 , there was a nasty issue in your code (swapped arguments in busTransfer).

**Comment by JulioCesarMatias** [comment]
@robustini Write and Read works with @hogthrob commit?

**Comment by JulioCesarMatias** [comment]
> @JulioCesarMatias ,  haven't had the opportunity to review your changes, but you may want to  have a look at my last commit https://github.com/iNavFlight/inav/commit/101e345b4f21afe2319c0235d0762082126a9d09 , there was a nasty issue in your code (swapped arguments in busTransfer).

Nice find

**Comment by JulioCesarMatias** [comment]
> @JulioCesarMatias ,  haven't had the opportunity to review your changes, but you may want to  have a look at my last commit https://github.com/iNavFlight/inav/commit/101e345b4f21afe2319c0235d0762082126a9d09 , there was a nasty issue in your code (swapped arguments in busTransfer).

why did you delete the "flashPartitionSet" function? In the last commit of my branch, I added the "flush" function

**Comment by hogthrob** [comment]
@JulioCesarMatias : I did not delete the flashPartitionSet function, I just temporarily disabled one call for creating the bad block management partition as I am not sure what this was for and because I wanted to have a single partition scheme to debug.

BTW, reading seems to work (read all 0xff) but writing is not working with my commit. I did not put your and my changes together yet.
I was trying to get the cli commands flash_read&flash_write to work (#define USE_FLASH_TOOLS) to have an easier way to debug things.  flash_read returns reasonable data now (that is how I spotted the swap problem) but flash_write does not write anything yet.

**Comment by hogthrob** [comment]
@JulioCesarMatias The last commit to my branch now provides working read and write. busTransferMultiple must be used when a command is combined with a data transfer to/from a buffer.

**Comment by JulioCesarMatias** [comment]
excellent @hogthrob, your effort was really good. Thank you very much for your contributions. Below I'll put the links of the .Hex files for you and Marco to test...

@robustini >> https://drive.google.com/file/d/1nTmH_UTORe-SNtR_5wQAqs5rB0xTWAwl/view?usp=sharing

@hogthrob >> https://drive.google.com/file/d/1C9oG-B0BhYctCiJcJC3x2HACm12Mt-Td/view?usp=sharing

**Comment by hogthrob** [comment]
@JulioCesarMatias : Couldn't test your hex file. I have a MAMBAH743, not a MATEKH743. Not sure if your hex file is for a MAMBA. Anyways, used your branch to build a MAMBAH743 and unfortunately, this image is not booting correctly at all (it is not giving any sign of life, USB VCP not detected).

**Comment by hogthrob** [comment]
@JulioCesarMatias I now build the origin/master branch, no problems, boots as expected. However, your branch does not work. I checked the differences and tried to find the magic difference. So far no luck.

**Comment by hogthrob** [comment]
@JulioCesarMatias Just saw your message. The problem is clearly somewhere in the flash_w25n01g code, as disabling USE_FLASH_W25N01G makes my board boot again.

**Comment by JulioCesarMatias** [comment]
> @JulioCesarMatias Just saw your message. The problem is clearly somewhere in the flash_w25n01g code, as disabling USE_FLASH_W25N01G makes my board boot again.

could it be that the flush function I added is what is causing the problem?

**Comment by hogthrob** [comment]
No, it is the `flashPartitionSet(FLASH_PARTITION_TYPE_BADBLOCK_MANAGEMENT...`   in flash_w25n01g.c

I commented out in my code, you removed the comment. Adding the comments again makes the code work on my machine. 
Can't tell you why, this needs further analysis.

**Comment by JulioCesarMatias** [comment]
ok, I generated a .Hex with this part of the code commented out again... Can you test it? >> https://drive.google.com/file/d/1wNRlb971dj0kx57hv5Mb2XLU4gRT07bU/view?usp=sharing

**Comment by hogthrob** [comment]
@JulioCesarMatias : No, I haven't. But I have now a "working" setup. Write / Erase seem to work. But I found that w25n01g_readBytes needs 2 attempts to read data correctly from if page changes.   
Tomorrow I will investigate what causes this.

**Comment by hogthrob** [comment]
The waitForReady in the flash_w25n01g.c approach / code is broken (was not checking the busy flag in most cases), which in turn breaks the read.
I have a fix which looks good for read at least, but a bit more investigation/thinking has to be done.  
I pushed the changes to my branch. Please DO NOT merge this changes yet as I intended to clean this up. Feel free to build and test it though.

**Comment by hogthrob** [comment]
Now it is time to test this.

**Comment by robustini** [comment]
Fixed with PR #8166.
